In [1]:
# Import basic packages

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
dataset=pd.read_csv('Churn_Modelling.csv')

In [3]:
dataset

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

# Data Preprocessing

In [4]:
# Drop the unwanted variables

dataset=dataset.iloc[:,3:]

In [5]:
dataset

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619    France  Female   42       2       0.00              1   
1             608     Spain  Female   41       1   83807.86              1   
2             502    France  Female   42       8  159660.80              3   
3             699    France  Female   39       1       0.00              2   
4             850     Spain  Female   43       2  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9995          771    France    Male   39       5       0.00              2   
9996          516    France    Male   35      10   57369.61              1   
9997          709    France  Female   36       7       0.00              1   
9998          772   Germany    Male   42       3   75075.31              2   
9999          792    France  Female   28       4  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0             1               1        101348.88       1  
1             0               1        112542.58       0  
2             1               0        113931.57       1  
3             0               0         93826.63       0  
4             1               1         79084.10       0  
...         ...             ...              ...     ...  
9995          1               0         96270.64       0  
9996          1               1        101699.77       0  
9997          0               1         42085.58       1  
9998          1               0         92888.52       1  
9999          1               0         38190.78       0  

[10000 rows x 11 columns]

#### Missing Values

In [6]:
dataset.isnull().sum()

CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

No missing values

#### Label Encoding

In [7]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(7), object(2)
memory usage: 859.5+ KB


In [8]:
# Label encoding Geography & Gender column

dataset['Geography']=dataset['Geography'].astype('category')
dataset['Geography']=dataset['Geography'].cat.codes

dataset['Gender']=dataset['Gender'].astype('category')
dataset['Gender']=dataset['Gender'].cat.codes

In [9]:
dataset

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619          0       0   42       2       0.00              1   
1             608          2       0   41       1   83807.86              1   
2             502          0       0   42       8  159660.80              3   
3             699          0       0   39       1       0.00              2   
4             850          2       0   43       2  125510.82              1   
...           ...        ...     ...  ...     ...        ...            ...   
9995          771          0       1   39       5       0.00              2   
9996          516          0       1   35      10   57369.61              1   
9997          709          0       0   36       7       0.00              1   
9998          772          1       1   42       3   75075.31              2   
9999          792          0       0   28       4  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0             1               1        101348.88       1  
1             0               1        112542.58       0  
2             1               0        113931.57       1  
3             0               0         93826.63       0  
4             1               1         79084.10       0  
...         ...             ...              ...     ...  
9995          1               0         96270.64       0  
9996          1               1        101699.77       0  
9997          0               1         42085.58       1  
9998          1               0         92888.52       1  
9999          1               0         38190.78       0  

[10000 rows x 11 columns]

In [10]:
# One hot encoding 

dataset=pd.get_dummies(dataset,columns=['Geography','Gender'])

In [11]:
dataset

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  Geography_0  Geography_1  \
0                  1        101348.88       1            1            0   
1                  1        112542.58       0            0            0   
2                  0        113931.57       1            1            0   
3                  0         93826.63       0            1            0   
4                  1         79084.10       0            0            0   
...              ...              ...     ...          ...          ...   
9995               0         96270.64       0            1            0   
9996               1        101699.77       0            1            0   
9997               1         42085.58       1            1            0   
9998               0         92888.52       1            0            1   
9999               0         38190.78       0            1            0   

      Geography_2  Gender_0  Gender_1  
0               0         1         0  
1               1         1         0  
2               0         1         0  
3               0         1         0  
4               1         1         0  
...           ...       ...       ...  
9995            0         0         1  
9996            0         0         1  
9997            0         1         0  
9998            0         0         1  
9999            0         1         0  

[10000 rows x 14 columns]

In [12]:
# drop the dummy variable Geography_0,Gender_0
dataset=dataset.drop(['Geography_0','Gender_0'],axis=1)

In [13]:
dataset

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  Geography_1  Geography_2  \
0                  1        101348.88       1            0            0   
1                  1        112542.58       0            0            1   
2                  0        113931.57       1            0            0   
3                  0         93826.63       0            0            0   
4                  1         79084.10       0            0            1   
...              ...              ...     ...          ...          ...   
9995               0         96270.64       0            0            0   
9996               1        101699.77       0            0            0   
9997               1         42085.58       1            0            0   
9998               0         92888.52       1            1            0   
9999               0         38190.78       0            0            0   

      Gender_1  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
9995         1  
9996         1  
9997         0  
9998         1  
9999         0  

[10000 rows x 12 columns]

#### Outlier

No need to remove outliers in classification problem

#### Feature scaling

Since we are using RF,Bagging method this step is not required

#### Check Imbalance dataset

In [14]:
dataset['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

It is a imbalance data

In [15]:
# Split the data into independent & depedent variables
x=dataset.drop(['Exited'],axis=1)
y=dataset['Exited']

In [16]:
x

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Geography_1  Geography_2  Gender_1  
0                  1        101348.88            0            0         0  
1                  1        112542.58            0            1         0  
2                  0        113931.57            0            0         0  
3                  0         93826.63            0            0         0  
4                  1         79084.10            0            1         0  
...              ...              ...          ...          ...       ...  
9995               0         96270.64            0            0         1  
9996               1        101699.77            0            0         1  
9997               1         42085.58            0            0         0  
9998               0         92888.52            1            0         1  
9999               0         38190.78            0            0         0  

[10000 rows x 11 columns]

In [17]:
y

0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64

In [18]:
# Since we do have imbalance dataset o we have to use oversmapling method to make our data balance

import imblearn
# status of imbalance
#method-oversampling
#majority class=0=7963
#minority class=1=2037

# status of balance
#class=0=7963
#class=1=7963

from imblearn.over_sampling import RandomOverSampler
over=RandomOverSampler()
x_over,y_over=over.fit_resample(x,y)

In [19]:
print(x_over.shape)
print(y_over.shape)

(15926, 11)
(15926,)


# Split the data into training & test for model building & evaluation

In [20]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x_over,y_over,train_size=0.75,random_state=30)

print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(11944, 11)
(3982, 11)
(11944,)
(3982,)


# Model 1- Bagging 

In [21]:
#Fit the model

from sklearn.ensemble import BaggingClassifier
bagging=BaggingClassifier()
bagging.fit(x_train,y_train)

BaggingClassifier()

In [22]:
y_pred_train=bagging.predict(x_train)
y_pred_test=bagging.predict(x_test)

In [23]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

#### Classification Report

In [24]:
print(classification_report(y_train,y_pred_train))
print("*"*80)
print(classification_report(y_test,y_pred_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      5983
           1       1.00      1.00      1.00      5961

    accuracy                           1.00     11944
   macro avg       1.00      1.00      1.00     11944
weighted avg       1.00      1.00      1.00     11944

********************************************************************************
              precision    recall  f1-score   support

           0       0.96      0.90      0.93      1980
           1       0.91      0.96      0.93      2002

    accuracy                           0.93      3982
   macro avg       0.93      0.93      0.93      3982
weighted avg       0.93      0.93      0.93      3982



#### Confusion Matrix

In [25]:
print(confusion_matrix(y_train,y_pred_train))
print("*"*30)
print(confusion_matrix(y_test,y_pred_test))

[[5965   18]
 [  20 5941]]
******************************
[[1780  200]
 [  76 1926]]


#### Accuracy score

In [26]:
print(accuracy_score(y_train,y_pred_train))
print("*"*20)
print(accuracy_score(y_test,y_pred_test))

0.9968184862692565
********************
0.9306880964339528


 There is no variance & bias

# Model 2- Random Forest

In [32]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier()
rf.fit(x_train,y_train)

RandomForestClassifier()

In [33]:
y_pred_rf_train=rf.predict(x_train)
y_pred_rf_test=rf.predict(x_test)

#### Classification Report

In [34]:
print(classification_report(y_train,y_pred_rf_train))
print("*"*80)
print(classification_report(y_test,y_pred_rf_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      5983
           1       1.00      1.00      1.00      5961

    accuracy                           1.00     11944
   macro avg       1.00      1.00      1.00     11944
weighted avg       1.00      1.00      1.00     11944

********************************************************************************
              precision    recall  f1-score   support

           0       0.98      0.91      0.94      1980
           1       0.91      0.98      0.94      2002

    accuracy                           0.94      3982
   macro avg       0.95      0.94      0.94      3982
weighted avg       0.94      0.94      0.94      3982



#### Confusion Matrix

In [35]:
print(confusion_matrix(y_train,y_pred_rf_train))
print("*"*30)
print(confusion_matrix(y_test,y_pred_rf_test))

[[5983    0]
 [   0 5961]]
******************************
[[1796  184]
 [  44 1958]]


#### Accuracy score

In [36]:
print(accuracy_score(y_train,y_pred_rf_train))
print("*"*20)
print(accuracy_score(y_test,y_pred_rf_test))

1.0
********************
0.9427423405323958


# Model 3- Decision Tree Classifier

In [37]:
from sklearn.tree import DecisionTreeClassifier
dt=DecisionTreeClassifier()
dt.fit(x_train,y_train)

DecisionTreeClassifier()

In [38]:
y_pred_dt_train=dt.predict(x_train)
y_pred_dt_test=dt.predict(x_test)

#### Classification Report

In [39]:
print(classification_report(y_train,y_pred_dt_train))
print("*"*80)
print(classification_report(y_test,y_pred_dt_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      5983
           1       1.00      1.00      1.00      5961

    accuracy                           1.00     11944
   macro avg       1.00      1.00      1.00     11944
weighted avg       1.00      1.00      1.00     11944

********************************************************************************
              precision    recall  f1-score   support

           0       0.98      0.85      0.91      1980
           1       0.87      0.98      0.92      2002

    accuracy                           0.91      3982
   macro avg       0.92      0.91      0.91      3982
weighted avg       0.92      0.91      0.91      3982



#### Confusion Matrix

In [40]:
print(confusion_matrix(y_train,y_pred_dt_train))
print("*"*30)
print(confusion_matrix(y_test,y_pred_dt_test))

[[5983    0]
 [   0 5961]]
******************************
[[1682  298]
 [  43 1959]]


#### Accuracy score

In [41]:
print(accuracy_score(y_train,y_pred_dt_train))
print("*"*20)
print(accuracy_score(y_test,y_pred_dt_test))

1.0
********************
0.914364640883978


# Model 4- Logistic Regression

In [42]:
from sklearn.linear_model import LogisticRegression
lr=LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

In [43]:
y_pred_lr_train=lr.predict(x_train)
y_pred_lr_test=lr.predict(x_test)

#### Classification Report

In [44]:
print(classification_report(y_train,y_pred_lr_train))
print("*"*80)
print(classification_report(y_test,y_pred_lr_test))

              precision    recall  f1-score   support

           0       0.66      0.65      0.66      5983
           1       0.66      0.66      0.66      5961

    accuracy                           0.66     11944
   macro avg       0.66      0.66      0.66     11944
weighted avg       0.66      0.66      0.66     11944

********************************************************************************
              precision    recall  f1-score   support

           0       0.66      0.65      0.65      1980
           1       0.66      0.66      0.66      2002

    accuracy                           0.66      3982
   macro avg       0.66      0.66      0.66      3982
weighted avg       0.66      0.66      0.66      3982



#### Confusion Matrix

In [45]:
print(confusion_matrix(y_train,y_pred_lr_train))
print("*"*30)
print(confusion_matrix(y_test,y_pred_lr_test))

[[3914 2069]
 [2027 3934]]
******************************
[[1286  694]
 [ 671 1331]]


#### Accuracy score

In [46]:
print(accuracy_score(y_train,y_pred_lr_train))
print("*"*20)
print(accuracy_score(y_test,y_pred_lr_test))

0.6570663094440723
********************
0.6572074334505273


There is High Bias

# Conclusion

Random Forest is the best model with accuarcy 94.27%